features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -0,0 +1,1087 @@
+    }
+
+    consume_char(_delimiter);
+
+    if (consume_char(start_tag)) {
+        _prefix_matched = true;
+        set_scope(ElemType);
+        return true;
+    }
+
+    // On mismatch go to information response scope
+    set_scope(InfoType);
+    return false;
+}
+
+bool ATHandler::consume_char(char ch)
+{
+    tr_debug("%s: %c", __func__, ch);
+    int read_char = get_char();
+    // If we read something else than ch, recover it
+    if (read_char != ch && read_char != -1) {
+        _recv_pos--;
+        return false;
+    }
+    return true;
+}
+
+bool ATHandler::consume_to_tag(const char *tag, bool consume_tag)
+{
+    tr_debug("%s: %s", __func__, tag);
+    size_t match_pos = 0;
+
+    while (true) {
+        int c = get_char();
+        if (c == -1) {
+            break;
+        } else if (c == tag[match_pos]) {
+            match_pos++;
+            if (match_pos == strlen(tag)) {
+                if (!consume_tag) {
+                    _recv_pos -= strlen(tag);
+                }
+                return true;
+            }
+        } else if (match_pos) {
+            match_pos = 0;
+        }
+    }
+    tr_error("consume_to_tag not found");
+    return false;
+}
+
+bool ATHandler::consume_to_stop_tag()
+{
+    tr_debug("%s", __func__);
+
+    if (!_stop_tag || (_stop_tag && _stop_tag->found) || _error_found) {
+        return true;
+    }
+
+    if (consume_to_tag((const char*)_stop_tag->tag, true)) {
+        return true;
+    }
+
+    tr_error("consume_to_stop_tag not found");
+    set_error(NSAPI_ERROR_DEVICE_ERROR);
+    return false;
+}
+
+// consume by size needed?
+
+void ATHandler::resp_stop()
+{
+    // Do not return on error so that we can consume whatever there is in the buffer
+
+    tr_debug("%s", __func__);
+
+    if (_current_scope == ElemType) {
+        information_response_element_stop();
+        set_scope(InfoType);
+    }
+
+    if (_current_scope == InfoType) {
+        information_response_stop();
+    }
+
+    // Go for response stop_tag
+    if (consume_to_stop_tag()) {
+        set_scope(NotSet);
+    }
+
+    // Restore stop tag to OK
+    set_tag(&_resp_stop, OK);
+    // Reset info resp prefix
+    memset(_info_resp_prefix, 0, sizeof(_info_resp_prefix));
+}
+
+void ATHandler::information_response_stop()
+{
+    tr_debug("%s", __func__);
+    if (consume_to_stop_tag()) {
+        set_scope(RespType);
+    }
+}
+
+void ATHandler::information_response_element_stop()
+{
+    tr_debug("%s", __func__);
+    if (consume_to_stop_tag()) {
+        set_scope(InfoType);
+    }
+}
+
+ATHandler::ScopeType ATHandler::get_scope()
+{
+    return _current_scope;
+}
+
+void ATHandler::set_string(char *dest, const char *src, size_t src_len)
+{
+    memcpy(dest, src, src_len);
+    dest[src_len] = '\0';
+}
+
+const char* ATHandler::mem_str(const char* dest, size_t dest_len, const char* src, size_t src_len)
+{
+    if (dest_len > src_len) {
+        for(size_t i = 0; i < dest_len-src_len; ++i) {
```

Comment:
Comparration stops too early. Fixed in https://github.com/ARMmbed/mbed-os/pull/6082/commits/49cffc23f8cc7cb63d4874440f729e4abc646f96